diff --git a/linux-user/elfload.c b/linux-user/elfload.c
index 5bccd2e2..8bee1371 100644
--- a/linux-user/elfload.c
+++ b/linux-user/elfload.c
@@ -8,6 +8,12 @@
 #include "disas/disas.h"
 #include "qemu/path.h"
 
+#include "../patches/qemu-sancov-common.h"
+
+abi_ulong sancov_entry_point;
+abi_ulong sancov_code_start;
+abi_ulong sancov_code_end;
+
 #ifdef _ARCH_PPC64
 #undef ARCH_DLINFO
 #undef ELF_PLATFORM
@@ -2300,6 +2306,18 @@ static void load_elf_image(const char *image_name, int image_fd,
     info->end_data = 0;
     info->brk = 0;
     info->elf_flags = ehdr->e_flags;
+    
+    if (!sancov_entry_point) {
+        char *ptr;
+        if ((ptr = getenv("QEMU_SANCOV_ENTRYPOINT")) != NULL)
+            sancov_entry_point = strtoul(ptr, NULL, 16);
+        else
+            sancov_entry_point = info->entry;
+#ifdef TARGET_ARM
+        /* The least significant bit indicates Thumb mode. */
+        sancov_entry_point &= ~(target_ulong)1;
+#endif
+    }
 
     for (i = 0; i < ehdr->e_phnum; i++) {
         struct elf_phdr *eppnt = phdr + i;
@@ -2335,9 +2353,11 @@ static void load_elf_image(const char *image_name, int image_fd,
             if (elf_prot & PROT_EXEC) {
                 if (vaddr < info->start_code) {
                     info->start_code = vaddr;
+                    if (!sancov_code_start) sancov_code_start = vaddr;
                 }
                 if (vaddr_ef > info->end_code) {
                     info->end_code = vaddr_ef;
+                    if (!sancov_code_end) sancov_code_end = vaddr_ef;
                 }
             }
             if (elf_prot & PROT_WRITE) {
@@ -2662,6 +2682,22 @@ int load_elf_binary(struct linux_binprm *bprm, struct image_info *info)
        change some of these later */
     bprm->p = setup_arg_pages(bprm, info);
 
+    // On PowerPC64 the entry point is the _function descriptor_
+    // of the entry function. For QEMU-SanCov to properly initialize,
+    // sancov_entry_point needs to be set to the actual first instruction
+    // as opposed executed by the target program. This as opposed to 
+    // where the function's descriptor sits in memory.
+    // copied from PPC init_thread
+#if defined(TARGET_PPC64) && !defined(TARGET_ABI32)
+    if (get_ppc64_abi(infop) < 2) {
+        uint64_t val;
+        get_user_u64(val, infop->entry + 8);
+        _regs->gpr[2] = val + infop->load_bias;
+        get_user_u64(val, infop->entry);
+        infop->entry = val + infop->load_bias;
+    }
+#endif
+
     scratch = g_new0(char, TARGET_PAGE_SIZE);
     if (STACK_GROWS_DOWN) {
         bprm->p = copy_elf_strings(1, &bprm->filename, scratch,
